### PR TITLE
[propolis] address cancel-safety issues with InstanceSerialConsoleHelper::recv

### PR DESCRIPTION
**Pull request by sunshowers**
TODO:

- [x] Expand on documentation.
- [x] Tests

See #434 and inline comments.

### CODE REVIEW

**Comment by sunshowers** [comment]
So sadly this means that while a migration is happening, the other select branches (and therefore progress) will be completely blocked. There is likely a way to write `recv()` such that other select branches aren't blocked. Going to give that a shot.

**Comment by lifning** [comment]
> So sadly this means that while a migration is happening, the other select branches (and therefore progress) will be completely blocked. There is likely a way to write `recv()` such that other select branches aren't blocked. Going to give that a shot.

I think it's actually okay to block the other select branch in this case - if a Text frame has been sent, it means the rest of the migration has completed already, and we shouldn't send characters any more until the InstanceSerialConsoleHelper's had its ws_stream replaced with the connection to the destination instance.

**Review by lifning — COMMENTED** [review]
WIP looks good so far. I'm currently imagining that tests might look like one demonstrating that putting a future that effectively does `x = ws.recv().process() => { ... }` in the `select!` causes the message to be lost outright, and one where `x = ws.recv() => { x.process() ... }` doesn't, where to avoid flaking we can force the cancellations to be predictable by having the 'destination server' be some mock that'll accept the `client.instance_serial()` call's connection and stall until we explicitly tell it to finish the handshake.

File: lib/propolis-client/src/lib.rs [hunk 1]
```diff
@@ -113,21 +113,50 @@ pub mod support {
...
-        /// Receive the next [WSMessage] from the server.
+        /// Receives the next [WSMessage] from the server, holding it in
+        /// abeyance until it is processed.
+        ///
         /// Returns [Option::None] if the connection has been terminated.
+        ///
+        /// # Cancel safety
+        ///
+        /// This method is cancel-safe and can be safely used in a `select!`
+        /// loop. However, to , and that portion is not cancel-safe.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply lib/propolis-client/src/lib.rs]: Done.

**Review by lifning — APPROVED** [review]

**Pull request merged**